akka-stream/src/main/scala/akka/stream/impl/UnfoldResourceSourceAsync.scala

```diff
@@ -27,83 +26,86 @@ import scala.util.control.NonFatal
   override def initialAttributes: Attributes = DefaultAttributes.unfoldResourceSourceAsync
 
   def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) with OutHandler {
-    lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
-    var resource = Promise[S]()
-    var open = false
-    implicit val context = ExecutionContexts.sameThreadExecutionContext
-
-    setHandler(out, this)
-
-    override def preStart(): Unit = createStream(false)
-
-    private def createStream(withPull: Boolean): Unit = {
-      val createdCallback = getAsyncCallback[Try[S]] {
-        case scala.util.Success(res) ⇒
-          open = true
-          resource.success(res)
-          if (withPull) onPull()
-        case scala.util.Failure(t) ⇒ failStage(t)
-      }
-      try {
-        create().onComplete(createdCallback.invoke)
-      } catch {
-        case NonFatal(ex) ⇒ failStage(ex)
-      }
-    }
-
-    private def onResourceReady(f: (S) ⇒ Unit): Unit = resource.future.foreach(f)
+    private lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+    private implicit def ec = ActorMaterializerHelper.downcast(materializer).system.dispatcher
```

Comment:
I'm not convinced that no future code would throw in this stage, so better not use the same-thread one